map.c

```diff
@@ -918,8 +1259,18 @@ static void *worker_pipeline(void *shared, int step, void *in)
 				s->buf[i] = mm_tbuf_init();
 #if defined(__AMD_SPLIT_KERNELS__)
 			s->trbuf = (mm_trbuf_t**)calloc(p->n_threads, sizeof(mm_trbuf_t*));
-			for (i = 0; i < p->n_threads; ++i)
-				s->trbuf[i] = mm_trbuf_init();
+			if (p->opt->flag & MM_F_GPU_CHAIN){
+				// TODO: make misc different for each ream
+				Misc misc = build_misc(s->p->mi, s->p->opt, 0, 1);
+				init_stream_gpu(&s->batch_max_anchors, &s->batch_max_reads, &s->gpu_min_n, misc);
+				// TODO: initialize GPU infrastructures
+				// TODO: need a data structure to remember all the memptrs 
+			} else {
```

Comment:
This is called mulitple time during an given execution, while `init_stream_gpu()` can only be called once until `finish_stream_gpu()` are call for each stream, which is call by each `worder_for` thread when i=-1 is fed. Any ideas how this works?

Reply:
My current version splits out error: `Error in gpu/plmem.cu:53 plmem_malloc_device_mem(): out of memory.` because this `init_stream_gpu` is called twice before any `finish_stream_gpu` is called.

Reply:
worker_for is called with -1 when there are no more reads to process for that thread.

Reply:
I understand that. But why is this part executed twice before any worker_for call with -1 is called?

Reply:
I am not familiar with build_misc.. is this in your standalone code?

Reply:
If in minimap2 (the reorganized one), this call with -1 should happen just once for each thread.

Reply:
I think I figured out why: there a three threads running worker_pipeline, each has input 0, 1, 2. Our original initialization function (TODO you left for GPU infrastructures initialization and trbuf initialization) was in step_0, which happens before step_1, but can run several times before step 1 for the first buf has finished. I move this trbuf and GPU initialization to step1 and I think I've fixed the problem here.

Reply:
I believe the TODOs were added by Xueshen. Need his input for understanding his thought process here. Step 0 in worker_pipeline is invoked first where it reads all the sequences into memory. Step 1 is the mapping, which we are further splitting into different stages (seed, chain and align) where we will call the seed function as we start processing reads, but will accumulate until we have a critical mass of reads before we send that batch for chain. All this is handled in worker_for. 

You are the expert in what your standalone requires and looks like you have it figured out. If you are unsure about anything, we can talk about it in our next sync.